src/backend/distributed/commands/create_citus_local_table.c

```diff
@@ -0,0 +1,620 @@
+	 * done in this function should be automic.
+	 */
+	LockRelationOid(shellRelationId, ExclusiveLock);
+
+	/*
+	 * Get necessary commands to recreate the shell table before renaming the
+	 * given relation to the shard relation.
+	 */
+	List *foreignConstraintCommands =
+		GetForeignConstraintCommandsTableInvolved(shellRelationId);
+
+	/*
+	 * Include DEFAULT clauses for columns getting their default values from
+	 * a sequence.
+	 */
+	bool includeSequenceDefaults = true;
+
+	List *shellTableDDLEvents = GetTableDDLEvents(shellRelationId,
+												  includeSequenceDefaults);
+	shellTableDDLEvents = list_concat(shellTableDDLEvents, foreignConstraintCommands);
+
+	uint64 shardId = GetNextShardId();
+
+	char *shellRelationName = get_rel_name(shellRelationId);
+	Oid shellRelationSchemaId = get_rel_namespace(shellRelationId);
+
+	/* below we convert relation with shellRelationId to the shard relation */
+	CreateCitusLocalTableShard(shellRelationId, shardId);
+
+	/*
+	 * As we retrieved the DDL commands necessary to create the shell table
+	 * from scratch, below we simply recreate the shell table executing them
+	 * via process utility.
+	 */
+	CreateShellTableForCitusLocalTable(shellTableDDLEvents);
+
+	/* update shellRelationId so it points to the shell table that we just created */
+	shellRelationId = get_relname_relid(shellRelationName, shellRelationSchemaId);
+
+	/* assert that we created the shell table properly in the same schema */
+	Assert(OidIsValid(shellRelationId));
+
+	InsertMetadataForCitusLocalTable(shellRelationId, shardId);
+
+	/* foreign tables does not support TRUNCATE trigger */
+	if (RegularTable(shellRelationId))
+	{
+		CreateTruncateTrigger(shellRelationId);
+	}
+
+	if (ShouldSyncTableMetadata(shellRelationId))
+	{
+		CreateTableMetadataOnWorkers(shellRelationId);
+	}
+
+	/*
+	 * We've a custom way of foreign key graph invalidation,
+	 * see InvalidateForeignKeyGraph().
+	 */
+	if (TableReferenced(shellRelationId) || TableReferencing(shellRelationId))
+	{
+		InvalidateForeignKeyGraph();
+	}
+}
+
+
+/*
+ * CreateCitusLocalTableShard creates the one and only shard of the citus
+ * local table lazily. That means, this function suffixes shardId to:
+ *  - relation name,
+ *  - all the objects "defined on" the relation and
+ *  - the foreign keys referencing to the relation.
+ */
+static void
+CreateCitusLocalTableShard(Oid relationId, uint64 shardId)
+{
+	RenameRelationToShardRelation(relationId, shardId);
+	RenameShardRelationConstraints(relationId, shardId);
+	RenameForeignConstraintsReferencingToShard(relationId, shardId);
+	RenameShardRelationIndexes(relationId, shardId);
+}
+
+
+/*
+ * RenameRelationToShardRelation appends given shardId to the end of the name
+ * of relation with shellRelationId.
+ */
+static void
+RenameRelationToShardRelation(Oid shellRelationId, uint64 shardId)
+{
+	Oid shellRelationSchemaId = get_rel_namespace(shellRelationId);
```

Comment:
Looking into the codebase, I guess we less prefer shortening words in variable names

Reply:
I would say that the more local a variable is, the shortened it can be, and as a developer if we see a short variable we will assume that they are local, for example if I see `i` I directly assume that it is a for loop variable not a function parameter etc.  But a variable should also be long enough so that it is not confusing and explanatory. 

Well these are probably okay because this is more of a general discussion, so I am okay in this PR with these variables.